docs/hoon/library/2b.md

```diff
@@ -4,28 +4,132 @@ navhome: /docs/
 
 
 # 2b: list logic
-Reverse
+### `++fand`
+
+All indices in `list`
+
+Produces the indices of all occurrences of `list` `nedl` in
+`list` `hstk` as a `list` of atoms.
+
+#### Accepts
+
+`nedl` is a list.
+
+`hstk` is a list.
+
+#### Produces
+
+A `list`.
+
+#### Source
+
+    ++  fand                                                ::  all indices
+      ~/  %fand
+      |=  [nedl=(list) hstk=(list)]
+      =|  i=@ud
+      =|  fnd=(list @ud)
+      |-  ^+  fnd
+      =+  [n=nedl h=hstk]
+      |-
+      ?:  |(?=($~ n) ?=($~ h))
+        (flop fnd)
+      ?:  =(i.n i.h)
+        ?~  t.n
+          ^$(i +(i), hstk +.hstk, fnd [i fnd])
+        $(n t.n, h t.h)
+      ^$(i +(i), hstk +.hstk)
+
+
+#### Examples
+
+    > (fand [3]~ ~[1 2 3])
+    ~[2]
+
+    > (fand [4]~ ~[1 2 3])
+    ~
+
+    > (fand ['a']~ "cbabab")
+    ~[2 4]
+
+    > (fand "ba" "cbabab")
+    ~[1 3]
+
```

Comment:
Is there any particular reason you used `[blah]~` and not `~[blah]` in these examples?  These are equivalent when `blah` is an atom but not otherwise.  My habit has been to prefer putting the `~` in front for lists, whereas putting it at the end ensures the result is always taken as a one-item list.  But I don't know where my preference came from.  To be clear, I'm not making a criticism, I'm just wondering about what the standard is.  @joemfb @belisarius222 @keatondunsford ?

Reply:
> ... putting `~` at the end ensures the result is always taken as a one-item list ...

Exactly. The difference between `~[blah]` and `~[blah]` is not due to `blah` being an atom -- it's just the number of hoons inside the brackets:

```
> [1 2]~
[[1 2] ~]
> ~[1 2]
[1 2 ~]
> =+(a=[1 2] ~[a])
[[1 2] ~]
> =+(a=[1 2] [a]~)
[[1 2] ~]
```

I don't have any particular preference for these examples. These expressions for null-terminated tuples are confusing at first, then pretty easy to remember.

Reply:
"The difference between ~[blah] and ~[blah] is not due to blah being an atom -- it's just the number of hoons inside the brackets:"

`[(add 12 (mul 2 3))]~`

This has multiple Hoons in it, but because it evaluates to an atom it's the same as:

`~[(add 12 (mul 2 3))]`

Reply:
You're correct that my initial criterion is wrong:

```
[[12 14]]~
```

is the same as

```
~[[12 14]]
```